File: Project_Bike_Rental.R
#Clean the environment
rm(list = ls())

#Set working directory
setwd("C:/Users/visu/Desktop/Edwisor/Projects/Project 1")

#Load the librarires
libraries = c("plyr","dplyr", "ggplot2","rpart","dplyr","DMwR","randomForest","usdm","corrgram","DataCombine")
lapply(X = libraries,require, character.only = TRUE)
rm(libraries)

#Read the csv file
day = read.csv(file = "day.csv", header = T, sep = ",", na.strings = c(" ", "", "NA"))

########################################EXPLORE THE DATA########################################
#First few rows
head(day)

#Dimensions of data
dim(day)

#Column names
names(day)

#Structure of variables
str(day)

########################################FEATURE ENGINEERING########################################
#Create columns
day$actual_temp <- day$temp*39
day$actual_feel_temp <- day$atemp*50
day$actual_windspeed <- day$windspeed*67
day$actual_hum = day$hum * 100

day$actual_season = factor(x = day$season, levels = c(1,2,3,4), labels = c("Spring","Summer","Fall","Winter"))
day$actual_yr = factor(x = day$yr, levels = c(0,1), labels = c("2011","2012"))
day$actual_holiday = factor(x = day$holiday, levels = c(0,1), labels = c("Working day","Holiday"))
day$actual_weathersit = factor(x = day$weathersit, levels = c(1,2,3,4), 
                               labels = c("Clear","Cloudy/Mist","Rain/Snow/Fog","Heavy Rain/Snow/Fog"))

day$weathersit = as.factor(day$weathersit)
day$season = as.factor(day$season)
day$dteday = as.character(day$dteday)
day$mnth = as.factor(day$mnth)
day$weekday = as.factor(as.character(day$weekday))
day$workingday = as.factor(as.character(day$workingday))
day$yr = as.factor(day$yr)
day$holiday = as.factor(day$holiday)

########################################MISSING VALUES########################################
missing_values = sapply(day, function(x){sum(is.na(x))})


########################################EXPLORE USING GRAPHS########################################
#Check the distribution of categorical Data using bar graph
bar1 = ggplot(data = day, aes(x = actual_season)) + geom_bar() + ggtitle("Count of Season")
bar2 = ggplot(data = day, aes(x = actual_weathersit)) + geom_bar() + ggtitle("Count of Weather")
bar3 = ggplot(data = day, aes(x = actual_holiday)) + geom_bar() + ggtitle("Count of Holiday")
bar4 = ggplot(data = day, aes(x = workingday)) + geom_bar() + ggtitle("Count of Working day")
# ## Plotting plots together
gridExtra::grid.arrange(bar1,bar2,bar3,bar4,ncol=2)

#Check the distribution of numerical data using histogram
hist1 = ggplot(data = day, aes(x =actual_temp)) + ggtitle("Distribution of Temperature") + geom_histogram(bins = 25)
hist2 = ggplot(data = day, aes(x =actual_hum)) + ggtitle("Distribution of Humidity") + geom_histogram(bins = 25)
hist3 = ggplot(data = day, aes(x =actual_feel_temp)) + ggtitle("Distribution of Feel Temperature") + geom_histogram(bins = 25)
hist4 = ggplot(data = day, aes(x =actual_windspeed)) + ggtitle("Distribution of Windspeed") + geom_histogram(bins = 25)
gridExtra::grid.arrange(hist1,hist2,hist3,hist4,ncol=2)

#Check the distribution of numerical data using scatterplot
scat1 = ggplot(data = day, aes(x =actual_temp, y = cnt)) + ggtitle("Distribution of Temperature") + geom_point() + xlab("Temperature") + ylab("Bike COunt")
scat2 = ggplot(data = day, aes(x =actual_hum, y = cnt)) + ggtitle("Distribution of Humidity") + geom_point(color="red") + xlab("Humidity") + ylab("Bike COunt")
scat3 = ggplot(data = day, aes(x =actual_feel_temp, y = cnt)) + ggtitle("Distribution of Feel Temperature") + geom_point() + xlab("Feel Temperature") + ylab("Bike COunt")
scat4 = ggplot(data = day, aes(x =actual_windspeed, y = cnt)) + ggtitle("Distribution of Windspeed") + geom_point(color="red") + xlab("Windspeed") + ylab("Bike COunt")
gridExtra::grid.arrange(scat1,scat2,scat3,scat4,ncol=2)

#Check for outliers in data using boxplot
cnames = colnames(day[,c("actual_temp","actual_feel_temp","actual_windspeed","actual_hum")])
for (i in 1:length(cnames))
{
  assign(paste0("gn",i), ggplot(aes_string(y = cnames[i]), data = day)+
           stat_boxplot(geom = "errorbar", width = 0.5) +
           geom_boxplot(outlier.colour="red", fill = "grey" ,outlier.shape=18,
                        outlier.size=1, notch=FALSE) +
           theme(legend.position="bottom")+
           labs(y=cnames[i])+
           ggtitle(paste("Box plot for",cnames[i])))
}
gridExtra::grid.arrange(gn1,gn3,gn2,gn4,ncol=2)

#Remove outliers in Windspeed
val = day[,19][day[,19] %in% boxplot.stats(day[,19])$out]
day = day[which(!day[,19] %in% val),]

#Check for multicollinearity using VIF
df = day[,c("instant","temp","atemp","hum","windspeed")]
vifcor(df)

#Check for collinearity using corelation graph
corrgram(day, order = F, upper.panel=panel.pie, text.panel=panel.txt, main = "Correlation Plot")

#Remove the unwanted variables

day <- subset(day, select = -c(holiday,instant,dteday,atemp,casual,registered,actual_temp,actual_feel_temp,actual_windspeed,
                               actual_hum,actual_season,actual_yr,actual_holiday,actual_weathersit))

rmExcept(keepers = "day")
########################################DECISION TREE########################################
#MAPE: 17.47%
#MAE: 684
#RMSE: 864.8
#Accuracy: 82.53%

#Divide the data into train and test
set.seed(123)
train_index = sample(1:nrow(day), 0.8 * nrow(day))
train = day[train_index,]
test = day[-train_index,]

#rpart for regression
dt_model = rpart(cnt ~ ., data = train, method = "anova")

#Predict the test cases
dt_predictions = predict(dt_model, test[,-10])

#Create dataframe for actual and predicted values
df = data.frame("actual"=test[,10], "pred"=dt_predictions)
head(df)

#calculate MAPE
regr.eval(trues = test[,10], preds = dt_predictions, stats = c("mae","mse","rmse","mape"))

#calculate MAPE
MAPE = function(actual, pred){
  print(mean(abs((actual - pred)/actual)) * 100)
}
MAPE(test[,10], dt_predictions)

########################################RANDOM FOREST########################################
#MAPE: 10.68%
#MAE: 392
#RMSE: 535
#Accuracy: 89.32%

#Train the data using random forest
rf_model = randomForest(cnt~., data = train, ntree = 500)

#Predict the test cases
rf_predictions = predict(rf_model, test[,-10])

#Create dataframe for actual and predicted values
df = cbind(df,rf_predictions)
head(df)

#Calculate MAPE
regr.eval(trues = test[,10], preds = rf_predictions, stats = c("mae","mse","rmse","mape"))
MAPE(test[,10], rf_predictions)

########################################LINEAR REGRESSION########################################
#MAPE: 12.17%
#RMSE: 673
#Accuracy: 87.83%
#MAE: 494
#Adjusted R squared: 0.8373
#F-statistic: 110.2

#Train the data using linear regression
lr_model = lm(formula = cnt~., data = train)

#Check the summary of the model
summary(lr_model)

#Predict the test cases
lr_predictions = predict(lr_model, test[,-10])

#Create dataframe for actual and predicted values
df = cbind(df,lr_predictions)
head(df)

#Calculate MAPE
regr.eval(trues = test[,10], preds = lr_predictions, stats = c("mae","mse","rmse","mape"))
MAPE(test[,10], lr_predictions)

#Plot a graph for actual vs predicted values
plot(test$cnt,type="l",lty=2,col="green")
lines(lr_predictions,col="blue")

#Predict a sample data
predict(lr_model,test[2,])

File: Project_Bike_Rental.py
# -*- coding: utf-8 -*-
"""
Created on Thu Oct  3 12:16:06 2019

@author: pc
"""

#Import libraries
import os
import pandas as pd
import numpy as np
​
#import libraries for plots
import matplotlib.pyplot as plt
import seaborn as sns
% matplotlib inline  
#Set working directory
os.chdir("C:\\Users\\pc\\Desktop\\R\\projects\\Bike-Rental-master")
print(os.getcwd())
#Read the csv file
day = pd.read_csv("day.csv", sep=",")
#Get the number of rows and columns
day.shape
#Get first 5 rows
print(day.head())
​
#Get the data types of variables
print(day.dtypes)
#Create a new dataframe containing required columns and creating new columns
df = day.copy()
df.head()
​
#Create new columns
df['actual_temp'] = day['temp'] * 39
df['actual_feel_temp'] = day['atemp'] * 50
df['actual_windspeed'] = day['windspeed'] * 67
df['actual_hum'] = day['hum'] * 100
​
df['actual_season'] = day['season'].replace([1,2,3,4],["Spring","Summer","Fall","Winter"])
df['actual_yr'] = day['yr'].replace([0,1],["2011","2012"])
df['actual_holiday'] = day['holiday'].replace([0,1],["Working day","Holiday"])
df['actual_weathersit'] = day['weathersit'].replace([1,2,3,4],["Clear","Cloudy/Mist","Rain/Snow/Fog","Heavy Rain/Snow/Fog"])
#Check the data types od variables
df.dtypes
​
#Change the data types
df['weathersit'] = df['weathersit'].astype('category')
df['holiday'] = df['holiday'].astype('category')
df['yr'] = df['yr'].astype('category')
df['season'] = df['season'].astype('category')
df['workingday'] = df['workingday'].astype('category')
df['weekday'] = df['weekday'].astype('category')
df['mnth'] = df['mnth'].astype('category')
df['actual_season'] = df['actual_season'].astype('category')
df['actual_yr'] = df['actual_yr'].astype('category')
df['actual_holiday'] = df['actual_holiday'].astype('category')
df['actual_weathersit'] = df['actual_weathersit'].astype('category')
​
df.dtypes
#Check the count of values of categorical variables
print(df.workingday.value_counts())
print(df.weekday.value_counts())
print(df.mnth.value_counts())
print(df.actual_yr.value_counts())
print(df.actual_holiday.value_counts())
print(df.actual_weathersit.value_counts())
#Check if there are missing values
df.isnull().sum()
#Check the bar graph of categorical Data using factorplot
sns.set_style("whitegrid")
sns.factorplot(data=df, x='actual_season', kind= 'count',size=4,aspect=2)
sns.factorplot(data=df, x='actual_weathersit', kind= 'count',size=4,aspect=2)
sns.factorplot(data=df, x='workingday', kind= 'count',size=4,aspect=2)
​
#Check the distribution of numerical data using histogram
plt.hist(data=df, x='actual_temp', bins='auto', label='Temperature')
plt.xlabel('Temperature in Celcius')
plt.title("Temperature Distribution")
#Check the distribution of numerical data using histogram
plt.hist(data=df, x='actual_hum', bins='auto', label='Temperature')
plt.xlabel('Humidity')
plt.title("Humidity Distribution")
#Check for outliers in data using boxplot
sns.boxplot(data=df[['actual_temp','actual_feel_temp','actual_windspeed','actual_hum']])
fig=plt.gcf()
fig.set_size_inches(8,8)
#Remove outliers in Humidity
q75, q25 = np.percentile(df['actual_hum'], [75 ,25])
print(q75,q25)
iqr = q75 - q25
print(iqr)
min = q25 - (iqr*1.5)
max = q75 + (iqr*1.5)
print(min)
print(max)
​
df = df.drop(df[df.iloc[:,19] < min].index)
df = df.drop(df[df.iloc[:,19] > max].index)
#Remove outliers in Windspeed
q75, q25 = np.percentile(df['actual_windspeed'], [75 ,25])
print(q75,q25)
iqr = q75 - q25
print(iqr)
min = q25 - (iqr*1.5)
max = q75 + (iqr*1.5)
print(min)
print(max)
​
df = df.drop(df[df.iloc[:,18] < min].index)
df = df.drop(df[df.iloc[:,18] > max].index)
#Check for collinearity using corelation matrix.
cor_mat= df[:].corr()
mask = np.array(cor_mat)
mask[np.tril_indices_from(mask)] = False
fig=plt.gcf()
fig.set_size_inches(30,12)
sns.heatmap(data=cor_mat,mask=mask,square=True,annot=True,cbar=True)
#Check the distribution of Temperature and Humdity against Bike rental count using scatter plot
fig, axs = plt.subplots(1,2, figsize=(15, 5), sharey=True)
axs[0].scatter(data=df, x='actual_temp', y='cnt')
axs[1].scatter(data=df, x='actual_hum', y='cnt', color = 'red')
fig.suptitle('Scatter plot for Temperature and Humidity')
plt.xlabel("Humidity")
plt.ylabel("Count of bikes")
#Check the distribution of Feel Temperature and Windspeed against Bike rental count using scatter plot
fig, axs = plt.subplots(1,2, figsize=(15, 5), sharey=True)
axs[0].scatter(data=df, x='actual_feel_temp', y='cnt')
axs[1].scatter(data=df, x='actual_windspeed', y='cnt', color = 'red')
fig.suptitle('Scatter plot for Feel Temperature and Windspeed')
plt.xlabel("Windspeed")
plt.ylabel("Count of bikes")
df = df.drop(columns=['holiday','instant','dteday','atemp','casual','registered','actual_temp','actual_feel_temp',
                      'actual_windspeed','actual_hum','actual_season','actual_yr','actual_holiday','actual_weathersit'])
DECISION TREE
#MAPE: 18.40%

#Accuracy: 81.60%

#Import Libraries for decision tree
from sklearn.cross_validation import train_test_split
from sklearn.tree import DecisionTreeRegressor
#Divide data into train and test
train,test = train_test_split(df, test_size = 0.2, random_state = 123)
#Train the model
dt_model = DecisionTreeRegressor(random_state=123).fit(train.iloc[:,0:9], train.iloc[:,9])
#Predict the results of test data
dt_predictions = dt_model.predict(test.iloc[:,0:9])
df_dt = pd.DataFrame({'actual': test.iloc[:,9], 'pred': dt_predictions})
df_dt.head()
#Function for Mean Absolute Percentage Error
def MAPE(y_actual,y_pred):
    mape = np.mean(np.abs((y_actual - y_pred)/y_actual))
    return mape
#Calculate MAPE for decision tree
MAPE(test.iloc[:,9],dt_predictions)
#MAPE: 18.40%
#Accuracy: 81.60%
Random Forest
#MAPE: 13.10%

#Accuracy:86.90%

#Import library for RandomForestRegressor
from sklearn.ensemble import RandomForestRegressor
#Train the model
rf_model = RandomForestRegressor(n_estimators=500,random_state=123).fit(train.iloc[:,0:9], train.iloc[:,9])
#Predict the results of test data
rf_predictions = rf_model.predict(test.iloc[:,0:9])
#Create a dataframe for actual values and predicted values
df_rf = pd.DataFrame({'actual': test.iloc[:,9], 'pred': rf_predictions})
df_rf.head()
#Calculate MAPE
MAPE(test.iloc[:,9],rf_predictions)
#MAPE: 13.10%
#Accuracy:86.90%
Linear Regression
#MAPE:17.07%

#Accuracy: 82.93%

#Adjusted r2: 0.852

#F-stat: 122.7

#import libraries for Linear regression
import statsmodels.api as sm
from sklearn.metrics import mean_squared_error
#Train the model
lr_model = sm.OLS(train.iloc[:,9].astype(float), train.iloc[:,0:9].astype(float)).fit()
#Check the summary of model
lr_model.summary()
#Predict the results of test data
lr_predictions = lr_model.predict(test.iloc[:,0:9])
##Create a dataframe for actual values and predicted values
df_lr = pd.DataFrame({'actual': test.iloc[:,9], 'pred': lr_predictions})
df_lr.head()
#Calclulate MAPE
MAPE(test.iloc[:,9],lr_predictions)
#MAPE:19.08%
#Accuracy: 81.92% 
#Adjusted r2: 0.967
#F-stat: 1852
#Create continuous data. Save target variable first
train_lr = train[['cnt','temp','hum','windspeed']]
test_lr = test[['cnt','temp','hum','windspeed']]
##Create dummies for categorical variables
cat_names = ["season", "yr", "mnth", "weekday", "workingday", "weathersit"]
​
for i in cat_names:
    temp1 = pd.get_dummies(train[i], prefix = i)
    temp2 = pd.get_dummies(test[i], prefix = i)
    train_lr = train_lr.join(temp1)
    test_lr = test_lr.join(temp2)
#Train the model
lr_model = sm.OLS(train_lr.iloc[:,0].astype(float), train_lr.iloc[:,1:34].astype(float)).fit()
#summary of model
lr_model.summary()
#Predict the results of test data
lr_predictions = lr_model.predict(test_lr.iloc[:,1:34])
##Create a dataframe for actual values and predicted values
df_lr = pd.DataFrame({'actual': test_lr.iloc[:,0], 'pred': lr_predictions})
df_lr.head()
#Calclulate MAPE
MAPE(test_lr.iloc[:,0],lr_predictions)
#MAPE:17.07%
#Accuracy: 82.93%
#Adjusted r2: 0.852
#F-stat: 122.7
​
